### PR TITLE
Forests per host

### PR DESCRIPTION
**Pull request by dmcassel**
Implemented the forests-per-host setting. If PR is accepted, I'll write a wiki showing how to use it.

### CODE REVIEW

File: src/lib/modules/databases.js [hunk 1]
```diff
@@ -29,11 +29,62 @@ DBManager.databaseOperation = function(operation, database, callback) {
...
+    var count = dbSettings.forest['forests-per-host'];
+    // idenfity the hosts
+    var hosts = this.getConfiguration('hosts');
+    // build forest names
+    var forests = [], forestNames = [];
+    hosts.forEach(function(host, index) {
+        var hostSettings = common.objectSettings('hosts/' + host, this.env);
+        console.log('host is: ' + hostSettings['host-name']);
+        for (var i = 0; i < count; i++) {
+            forests.push({'host': hostSettings['host-name'], 'forest-name': dbSettings['database-name'] + '-' + (index * count + i)});
```
Comment (grtjn) [inline src/lib/modules/databases.js]: While at this, you could apply the same naming convention that Roxy applies: dbname-hostnr-forestnr, where hostnr is zero-padded to 3 digits.. :)

File: src/bin/mlsound-wipe.js [hunk 2]
```diff
@@ -36,12 +36,14 @@ prompt.get({
...
             logger.warning('Waiting for server restart');
             setTimeout(function() {
-                dbManager.removeDatabase('content', undefined, function() {
-                    dbManager.removeDatabase('modules', undefined, function() {
-                        dbManager.removeForests('full', function() {
-                            dbManager.removeUsers(function() {
-                                dbManager.removeRoles(function() {
-                                    logger.info('Project wiped!');
+                dbManager.removeForests('content', 'full', function() {
+                    dbManager.removeForests('modules', 'full', function() {
```
Comment (grtjn) [inline src/bin/mlsound-wipe.js:40]: Does this work? Shouldn't you detach forests first, then delete the databases, then delete the forests? (that is how Roxy does it ;) Or maybe detach+delete forests, then delete databases, but detach of forests must come first..

Comment (dmcassel) [inline src/bin/mlsound-wipe.js:40]: The removeForests function detaches and deletes the forests.

**Comment by dmcassel** [comment]
Don't accept this PR yet; I sometimes see a timing problem (attempting database creation before MarkLogic recognizes that the forests are ready). Thought I'd cleared that up, but I'll give it another look.

File: src/bin/mlsound-wipe.js (hunk 2 above)
Comment (grtjn) [inline src/bin/mlsound-wipe.js:40]: Yes, I saw that, but you are calling removeDatabase first (you switched that around for some reason)..

Comment (dmcassel) [inline src/bin/mlsound-wipe.js:40]: No, I'm calling removeForests first. mlsound uses a callback mechanism, so this removes content forests, then modules forests, then the content database, etc.

Comment (grtjn) [inline src/bin/mlsound-wipe.js:40]: Ah, overlooked that. Writing it as a chain would make it easier to read.. :-P

Comment (dmcassel) [inline src/bin/mlsound-wipe.js:40]: My preference would be using Promises, but I went with the project's style.

**Comment by rjrudin** [comment]
Food for thought - in ml-app-deployer, I went with a command approach, where each command is responsible for some portion of the deployment - e.g. a command for creating databases, a command for creating users, etc. Each command defines an order that it executes in during a deploy and during an undeploy (e.g. during deploy, a database has to be created before a server, and during an undeploy, a server has to be deleted before the database is deleted). 

The deployer object is then unaware of what any command does - he's just given a list of commands to execute, and he sorts them (based on whether he's doing a deploy or undeploy), and then executes them. This gives a nice open/closed effect where the deployer can be modified by adding new commands, and the order can be precisely controlled, all without modifying the deployer. 

I think the equivalent here would be that instead of wipe.js having explicit knowledge of everything that must be done and in what order, it would instead load up a list of commands (or have them injected somehow), sort them, and then execute them. Just something to consider.

**Comment by dmcassel** [comment]
Updated code to properly handle asynchronous checking for and creation of forests. @miguelrgonzalez, how does this look to you?

File: src/bin/mlsound-wipe.js (hunk 2 above)
Comment (miguelrgonzalez) [inline src/bin/mlsound-wipe.js:40]: I was debating which of the approaches to pick, and ended up with callbacks...
To be completely honest I'm not 100% happy with it neither, and chances are I will eventually move it to promises.

**Comment by miguelrgonzalez** [comment]
Hi,
thanks for the pull!!

It does look to me, only a couple of things:
- I see some console.log, please use the logger instead. There is some "nice" formatting around the output.
- Instead of using 

```
    --outStandingRequests; 
    checkFinished();
```

consider using callBackwhenDone() like the rest of the methods. It's functionally the same, but just for the sake of consistency.
- When splitting functionality in modules, I thought that it made sense to have forest operations under servers.js (that's were the remove operations are), but I see you placed buildForestsByHost function under databases. This function should be moved to keep things consistent.

Again, thanks a lot for expending time on this!
Cheers,
Miguel

**Comment by dmcassel** [comment]
@miguelrgonzalez, updated as specified.

**Comment by miguelrgonzalez** [comment]
@dmcassel Looks like accepting your other 2 pull requests, created some conflicts...

**Comment by dmcassel** [comment]
@miguelrgonzalez Conflicts fixed

**Comment by miguelrgonzalez** [comment]
merged!

**Pull request merged**